File: server/src/protocol.ts
import { z } from "zod";

/**
 * ワイヤプロトコルの定義。ここが client / server 間の契約の正。
 *
 * - client → server: `clientMessageSchema`
 * - server → client: `serverEventSchema`（events テーブルの1行がそのまま配送単位）
 *   例外として、永続化しない一時フレーム `transientErrorSchema` がある
 */

export const clientMessageSchema = z.object({
  type: z.literal("user_message"),
  payload: z.object({
    text: z.string().min(1).max(8_000),
  }),
});

export type ClientMessage = z.infer<typeof clientMessageSchema>;

export const eventTypeSchema = z.enum(["user_message", "assistant_message", "notification"]);

export type EventType = z.infer<typeof eventTypeSchema>;

export const serverEventSchema = z.object({
  id: z.number().int().positive(),
  type: eventTypeSchema,
  ts: z.string(),
  payload: z.object({
    text: z.string(),
  }),
});

export type ServerEvent = z.infer<typeof serverEventSchema>;

/** 検証エラー等、履歴に残す価値のない一時フレーム（id を持たない） */
export const transientErrorSchema = z.object({
  type: z.literal("error"),
  payload: z.object({
    message: z.string(),
  }),
});

export type TransientError = z.infer<typeof transientErrorSchema>;

/**
 * アシスタント応答のストリーミング断片。永続化せず、接続中のクライアントにのみ流す。
 * 全文は最後に assistant_message イベントとして受信箱に載るため、
 * 取りこぼしても catch-up で完全な形が手に入る（ADR-0003 の原則そのまま）
 */
export const transientDeltaSchema = z.object({
  type: z.literal("assistant_delta"),
  payload: z.object({
    text: z.string(),
  }),
});

export type TransientDelta = z.infer<typeof transientDeltaSchema>;

export function parseClientMessage(raw: string): ClientMessage | null {
  let json: unknown;
  try {
    json = JSON.parse(raw);
  } catch {
    return null;
  }
  const result = clientMessageSchema.safeParse(json);
  return result.success ? result.data : null;
}

File: app/lib/protocol.dart
import 'dart:convert';

/// サーバー発イベント。server/src/protocol.ts の serverEventSchema と対応する。
class ServerEvent {
  const ServerEvent({
    required this.id,
    required this.type,
    required this.ts,
    required this.text,
  });

  final int id;
  final String type;
  final String ts;
  final String text;

  bool get isUser => type == 'user_message';
  bool get isNotification => type == 'notification';

  /// 1 WebSocket フレーム / catch-up の1要素をパースする。
  /// 永続イベントでないもの（一時 error 等）や未知の形は null。
  static ServerEvent? tryParse(Object? raw) {
    final Object? json;
    if (raw is String) {
      try {
        json = jsonDecode(raw);
      } on FormatException {
        return null;
      }
    } else {
      json = raw;
    }
    if (json is! Map<String, dynamic>) return null;
    final id = json['id'];
    final type = json['type'];
    final ts = json['ts'];
    final payload = json['payload'];
    if (id is! int || type is! String || ts is! String) return null;
    if (payload is! Map<String, dynamic> || payload['text'] is! String) {
      return null;
    }
    return ServerEvent(
      id: id,
      type: type,
      ts: ts,
      text: payload['text'] as String,
    );
  }
}

String encodeUserMessage(String text) => jsonEncode({
      'type': 'user_message',
      'payload': {'text': text},
    });

/// assistant_delta（ストリーミング断片・非永続）の断片テキストを返す。違えば null。
String? tryParseAssistantDelta(Object? raw) {
  if (raw is! String) return null;
  final Object? json;
  try {
    json = jsonDecode(raw);
  } on FormatException {
    return null;
  }
  if (json is! Map<String, dynamic>) return null;
  if (json['type'] != 'assistant_delta') return null;
  final payload = json['payload'];
  if (payload is! Map<String, dynamic>) return null;
  final text = payload['text'];
  return text is String ? text : null;
}
